dotCMS/src/main/java/com/dotcms/rest/api/CorsFilter.java

```diff
@@ -1,23 +1,86 @@
 package com.dotcms.rest.api;
 
+import java.io.IOException;
+import java.io.StringWriter;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
 import com.dotcms.repackage.javax.ws.rs.container.ContainerRequestContext;
 import com.dotcms.repackage.javax.ws.rs.container.ContainerResponseContext;
 import com.dotcms.repackage.javax.ws.rs.container.ContainerResponseFilter;
 import com.dotcms.repackage.javax.ws.rs.core.MultivaluedMap;
-import java.io.IOException;
+import com.dotmarketing.util.Config;
+import com.google.common.collect.ImmutableList;
+
 
 /**
  * @author Geoff M. Granum
  */
 public class CorsFilter implements ContainerResponseFilter {
 
+
+    final private static String CORS_PREFIX = "api.cors";
+    final private static String CORS_DEFAULT = "default";
+    final private static Map<String, List<String[]>> headerMap = new ConcurrentHashMap<>();
+
+
     @Override
     public void filter(ContainerRequestContext requestContext, ContainerResponseContext responseContext) throws IOException {
-        MultivaluedMap<String, Object> headers = responseContext.getHeaders();
+        final MultivaluedMap<String, Object> responseHeaders = responseContext.getHeaders();
+        final String resource = requestContext.getUriInfo().getMatchedResources().get(0).getClass().getSimpleName().toLowerCase();
+        getHeaders(resource).forEach(header-> responseHeaders.add(header[0], header[1]));
+        
+    }
+
+
+    protected List<String[]> getHeaders(final String mapping) {
```

Comment:
Since the headers are coming from the config are such as inmutable, probably you can load the headerMap as an immutable Map on a static block or so when the class just started and just return it, instead of using a lazy loading strategy, in that case does not need the concurrentMap and the syncro block just a eager strategy :)